### PR TITLE
Break Mints/Spend/Tx operations into streams

### PR DESCRIPTION
**Pull request by micahriggan**
The goal of this is to eliminate the need to hold an entire block's worth of mints, spends, and txs in memory all at once.

### CODE REVIEW

File: packages/bitcore-node/src/models/transaction.ts [hunk 1]
```diff
@@ -16,12 +14,16 @@ import { Config } from '../services/config';
...
+import { Readable, Transform } from 'stream';
+import { Collection } from 'mongodb';
+
 export { ITransaction };
 
 const { onlyWalletEvents } = Config.get().services.event;
 function shouldFire(obj: { wallets?: Array<ObjectID> }) {
   return !onlyWalletEvents || (onlyWalletEvents && obj.wallets && obj.wallets.length > 0);
 }
+const MAX_BATCH_SIZE = 50000;
```
Comment (micahriggan) [inline packages/bitcore-node/src/models/transaction.ts]: Better batching with Transform

File: packages/bitcore-node/src/models/transaction.ts [hunk 2]
```diff
@@ -75,6 +79,119 @@ export type SpendOp = {
...
+const getUpdatedBatchIfMempool = (batch, height) =>
+  height >= SpentHeightIndicators.minimum ? batch : batch.map(op => TransactionStorage.toMempoolSafeUpsert(op, height));
+
+export class MempoolSafeTransform extends Transform {
+  constructor(private height: number) {
+    super({ objectMode: true });
+  }
+
+  async _transform(
+    coinBatch: Array<{ updateOne: { filter: any; update: { $set: any; $setOnInsert?: any } } }>,
```
Comment (micahriggan) [inline packages/bitcore-node/src/models/transaction.ts]: ugly

File: packages/bitcore-node/src/models/transaction.ts [hunk 3]
```diff
@@ -75,6 +79,119 @@ export type SpendOp = {
...
+
+export class MempoolTxEventTransform extends Transform {
+  constructor(private height: number) {
+    super({ objectMode: true });
+  }
+
+  _transform(txBatch: Array<TxOp>, _, done) {
+    if (this.height < SpentHeightIndicators.minimum) {
+      const eventPayload = txBatch
+        .map(op => ({ ...op.updateOne.update.$set, ...op.updateOne.filter, ...op.updateOne.update.$setOnInsert }))
```
Comment (micahriggan) [inline packages/bitcore-node/src/models/transaction.ts]: ugly

File: packages/bitcore-node/src/models/transaction.ts [hunk 4]
```diff
@@ -293,24 +380,86 @@ export class TransactionModel extends BaseTransaction<IBtcTransaction> {
             upsert: true,
             forceServerObjectId: true
           }
-        };
-      });
-      return txOps;
+        });
+
+        if (txBatch.length > MAX_BATCH_SIZE) {
```
Comment (micahriggan) [inline packages/bitcore-node/src/models/transaction.ts]: Remove in favor of a batching transform

File: packages/bitcore-node/src/models/transaction.ts [hunk 5]
```diff
@@ -376,83 +526,41 @@ export class TransactionModel extends BaseTransaction<IBtcTransaction> {
...
-            .filter(wallet => wallet.address === mintOp.updateOne.update.$set.address)
-            .map(wallet => wallet.wallet);
-          mintOp.updateOne.update.$set.wallets = transformedWallets;
-          delete mintOp.updateOne.update.$setOnInsert.wallets;
-          if (!Object.keys(mintOp.updateOne.update.$setOnInsert).length) {
-            delete mintOp.updateOne.update.$setOnInsert;
-          }
-          return mintOp;
-        });
+      if (mintBatch.length >= MAX_BATCH_SIZE) {
```
Comment (micahriggan) [inline packages/bitcore-node/src/models/transaction.ts]: replace with Batcher

File: packages/bitcore-node/src/models/transaction.ts [hunk 6]
```diff
@@ -465,10 +573,18 @@ export class TransactionModel extends BaseTransaction<IBtcTransaction> {
             update: { $set: { spentTxid: tx._hash || tx.hash, spentHeight: height } }
           }
         };
-        spendOps.push(updateQuery);
+        spendOpsBatch.push(updateQuery);
       }
+      if (spendOpsBatch.length > MAX_BATCH_SIZE) {
```
Comment (micahriggan) [inline packages/bitcore-node/src/models/transaction.ts]: replace with batcher

File: packages/bitcore-node/src/modules/bitcoin/p2p.ts [hunk 7]
```diff
@@ -192,7 +192,7 @@ export class BitcoinP2PWorker extends BaseP2PWorker<IBtcBlock> {
     logger.debug('Getting block, hash:', hash);
     let received = false;
     return new Promise<Bitcoin.Block>(async resolve => {
-      this.events.once(hash, (block: Bitcoin.Block) => {
+      this.events.once(hash, async (block: Bitcoin.Block) => {
```
Comment (micahriggan) [inline packages/bitcore-node/src/modules/bitcoin/p2p.ts]: remove

File: packages/bitcore-node/src/models/transaction.ts (hunk 4 above)
Comment (micahriggan) [reply packages/bitcore-node/src/models/transaction.ts]: Tried it, works better with batching in each function, due to the tagging

**Comment by msalcala11** [comment]
ACK 👍

**Comment by matiu** [comment]
Please add tests.

**Comment by matiu** [comment]
Great job! nice tests!

**Pull request merged**